cmd/geth/simulatedbeacon/simulated_beacon.go

```diff
@@ -0,0 +1,237 @@
+func (c *SimulatedBeacon) Start() error {
+	c.shutdownCh = make(chan struct{})
+	go c.loop()
+	return nil
+}
+
+// Stop halts the SimulatedBeacon service
+func (c *SimulatedBeacon) Stop() error {
+	close(c.shutdownCh)
+	return nil
+}
+
+// loop manages the lifecycle of the SimulatedBeacon.
+// it drives block production, taking the role of a CL client and interacting with Geth via public engine/eth APIs
+func (c *SimulatedBeacon) loop() {
+	var (
+		ticker             = time.NewTimer(time.Second * time.Duration(c.period))
+		buildWaitTime      = time.Millisecond * 100
+		header             = c.eth.BlockChain().CurrentHeader()
+		lastBlockTime      = header.Time
+		engineAPI          = catalyst.NewConsensusAPI(c.eth)
+		curForkchoiceState = engine.ForkchoiceStateV1{
+			HeadBlockHash:      header.Hash(),
+			SafeBlockHash:      header.Hash(),
+			FinalizedBlockHash: header.Hash(),
+		}
+	)
+	// if genesis block, send forkchoiceUpdated to trigger transition to PoS
+	if header.Number.Sign() == 0 {
+		if _, err := engineAPI.ForkchoiceUpdatedV2(curForkchoiceState, nil); err != nil {
+			log.Crit("failed to initiate PoS transition for genesis via Forkchoiceupdated", "err", err)
+		}
+	}
+
+	beginSealing := func() (engine.ForkChoiceResponse, error) {
+		tstamp := uint64(time.Now().Unix())
+		if tstamp <= lastBlockTime {
+			tstamp = lastBlockTime + 1
+		}
+		return engineAPI.ForkchoiceUpdatedV2(curForkchoiceState, &engine.PayloadAttributes{
+			Timestamp:             tstamp,
+			SuggestedFeeRecipient: c.getFeeRecipient(),
+			Withdrawals:           c.withdrawals.pop(),
+		})
+	}
+	finalizeSealing := func(id *engine.PayloadID, onDemand bool) error {
+		payload, err := engineAPI.GetPayloadV1(*id)
+		if err != nil {
+			return fmt.Errorf("error retrieving payload: %v", err)
+		}
+
+		if onDemand && len(payload.Transactions) == 0 && len(payload.Withdrawals) == 0 {
+			// If the payload is empty, despite there being pending transactions,
+			// that indicates that we need to give it more time to build the block.
+			if buildWaitTime < 10*time.Second {
+				buildWaitTime += buildWaitTime
+			}
+			// TODO: don't lose withdrawals
+			return nil
+		}
+		buildWaitTime = 100 * time.Millisecond // Reset it
+		// mark the payload as canon
+		if _, err = engineAPI.NewPayloadV2(*payload); err != nil {
+			return fmt.Errorf("failed to mark payload as canonical: %v", err)
+		}
+		curForkchoiceState = engine.ForkchoiceStateV1{
+			HeadBlockHash:      payload.BlockHash,
+			SafeBlockHash:      payload.BlockHash,
+			FinalizedBlockHash: payload.BlockHash,
+		}
+		// mark the block containing the payload as canonical
+		if _, err = engineAPI.ForkchoiceUpdatedV2(curForkchoiceState, nil); err != nil {
+			return fmt.Errorf("failed to mark block as canonical: %v", err)
+		}
+		lastBlockTime = payload.Timestamp
+		return nil
+	}
+	var fcId *engine.PayloadID
+	if fc, err := beginSealing(); err != nil {
+		log.Error("Error starting sealing-work", "err", err)
```

Comment:
Yes, I think `log.Crit` is usually taking the lazy way out... I find it more elegant to return errors and handle them, rather than log kill the process on failure. 
For example, if a shutdown were in progress, it would be acceptable if the engine rejected a block, due to shuttind down. And in that case, it's not "correct" to just "die". Better to log and exit gracefully, in general, IMO.